### PR TITLE
Improve top level auto open detection.

### PR DESCRIPTION
**Pull request by nojaf**
As we discussed in https://github.com/safesparrow/fsharp/pull/24#discussion_r1035365291.
This makes the last resort link to all files scenario stricter and pulls up the content in a module with a namespace prefix scenario.

### CODE REVIEW

File: tests/ParallelTypeCheckingTests/Code/AlwaysLinkDetection.fs [hunk 1]
```diff
@@ -35,13 +35,13 @@ let doesFileHasAutoOpenBehavior (ast: ParsedInput) : bool =
     match ast with
     | ParsedInput.SigFile (ParsedSigFileInput (contents = contents)) ->
         List.exists
-            (fun (SynModuleOrNamespaceSig (attribs = attribs; kind = kind)) ->
-                isAnyAttributeAutoOpen attribs
+            (fun (SynModuleOrNamespaceSig (attribs = attribs; longId = longId; kind = kind)) ->
+                (isAnyAttributeAutoOpen attribs && longId.Length < 2)
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Code/AlwaysLinkDetection.fs:39]: Rather than treat this specially, can we solve it in a generic way, by doing the same thing we do for nested modules?
These two are equivalent and could be solved in the same way:
```fsharp
[<AutoOpen>]
module A
...
```
```fsharp
namespace global
[<AutoOpen>]
module A =
  ...
```
In both cases we should be able to pull the contents of `A` up, which in this case means making the contents of `A` available through `::`/the root namespace, and through `A` - the only difference is that the parent is the root.

Doing that would make the code simpler.

Let me give it a go and see if I can adjust the code to do that.

File: tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs [hunk 2]
```diff
@@ -58,3 +58,59 @@ let ``Basic trie`` () =
...
+
+let a = 0
+"""
+                        )
+                }
+            |]
+
+    // Assert that both A and B expose file index 0
+    let aNode = trie.Children.["A"]
+    Assert.AreEqual(set [| 0 |], aNode.Files)
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs:85]: Why does `A` expose the file? There is nothing in the namespace `A` other than the nested module `A.B`, so just sharing the namespace `A` shouldn't add a link. I still think we can unify the top-level module and nested module handling and simplify the code.

File: tests/ParallelTypeCheckingTests/Code/AlwaysLinkDetection.fs (hunk 1 above)
Comment (nojaf) [reply tests/ParallelTypeCheckingTests/Code/AlwaysLinkDetection.fs:39]: There is a concept of `Root` right now in the codebase. It wouldn't be a stretch to add files to the root as well.

File: tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs (hunk 2 above)
Comment (nojaf) [reply tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs:85]: Opening `A` does give you access to the contents of `B`:

image

Comment (safesparrow) [reply tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs:85]: Yes, you're right, in this case, because of the let binding, we need a link.

However, the following test fails, and it should pass:
```fsharp
[<Test>]
let ``Toplevel AutoOpen module with multi prefixed namespace 2`` () =
    let trie =
        TrieMapping.mkTrie
            [|
                {
                    Idx = 0
                    File = "A.fsi"
                    AST =
                        parseSourceCode (
                            "A.fsi",
                            """
[<AutoOpen>]
module A.B.C

module D =
    let a = 0
"""
                        )
                }
            |]

    // Assert that only D exposes file index 0, because A, A.B, A.B.C contain only nested modules/namespaces.
    let aNode = trie.Children.["A"]
    Assert.AreEqual(Set.empty, aNode.Files)
    let bNode = aNode.Children.["B"]
    Assert.AreEqual(Set.empty, bNode.Files)
    let cNode = bNode.Children.["C"]
    Assert.AreEqual(Set.empty, cNode.Files)
    let dNode = cNode.Children.["D"]
    Assert.AreEqual(set [| 0 |], dNode.Files)
```

Comment (safesparrow) [reply tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs:85]: I think by treating top-level modules differently we're unnecessarily adding edge cases to handle and causing ourselves headaches. Best way would be for the code to treat an auto-open top-level module as if it were a nested module inside the global namespace.

File: tests/ParallelTypeCheckingTests/Code/TrieMapping.fs [hunk 3]
```diff
@@ -52,146 +52,164 @@ let rec mkTrieNodeFor (file: FileWithAST) : TrieNode =
...
-
-                            let children = List.choose (mkTrieForNestedSigModule idx) decls
-
-                            continuation (
-                                Dictionary<_, _>(
-                                    Seq.singleton (
-                                        KeyValuePair(
+        |> List.choose
+            (fun (SynModuleOrNamespaceSig (longId = longId; kind = kind; attribs = attribs; decls = decls; accessibility = _accessibility)) ->
+                let hasTypesOrAutoOpenNestedModules =
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Code/TrieMapping.fs:57]: I know this is already in the default branch, but:
- Do we only use this information for namespaces? What about modules that contain nested AutoOpen modules?
- Looks that we only check one nesting level - could we use the detection recursively?
- Looks that we treat AutoOpen nested modules the same way as types - I'd like to do something else.

The old solution, despite its flaws, at least at some point was doing the following:
- For a given file it would return a list of [nested] namespaces/modules defined in it
- If an AutoOpen module was found, that module segment would be 'skipped' and its contents would be linked to its parent. This wasn't technically correct, because you need to make it available with and without the segment.
So the only effect an AutoOpen had was that the returned implicit 'structure' of modules was slightly flattened.
Given that we now have a dedicated tree structure for this, we should be able to do the same, but this time correctly.

File: tests/ParallelTypeCheckingTests/Code/TrieMapping.fs [hunk 4]
```diff
@@ -52,146 +52,164 @@ let rec mkTrieNodeFor (file: FileWithAST) : TrieNode =
...
-                                                Current = current
-                                                Children = Dictionary(children)
-                                            }
+                                            (if hasTypesOrAutoOpenNestedModules then
+                                                 hs idx
+                                             else
+                                                 emptyHS ())
+                                        )
+                                    else
+                                        TrieNodeInfo.Module(name, idx)
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Code/TrieMapping.fs:96]: Does that mean that when a nested module is referenced, it always produces a link? Do we store the information about the contents of the module somewhere?

**Comment by safesparrow** [comment]
I added some comments, but I think there is a more general issue we're having, which is communication 😅 

I'm realising that many of my comments are the result of different assumptions about what the current algorithm aims to do.

For example, I was always convinced that we would end up with a solution that avoids a link if the only reference is an `open A.B` statement and `A.B` only contains nested modules.
But, as we discussed earlier, this would require a separate piece of information to be tracked, which we would pass into the TcState for each file, so that type-checking doesn't fail on that `open A.B` statement.

I think we should have a chat about that and other things. It's probable that the above 'optimisation' would be negligible, because how often does someone reference a module but not use its contents? Maybe not too often.
That would somewhat justify different handling of namespaces and modules.

Overall, I just don't understand the dependency resolution algorithm TBH. Maybe that's fine, but it prevents me from reasoning about it.
I guess we just switched the roles 😅

**Comment by safesparrow** [comment]
Unrelated to this PR, but...
Given that our `TrieNodeInfo` allows only one file for a module, how do we handle a situation where the same module is (incorrectly) defined in two places? I guess we will somewhat randomly pick one of the files as a dependency when referencing it?
Or does the code throw when trying to merge two modules with the same name, from different files?

File: tests/ParallelTypeCheckingTests/Code/TrieMapping.fs (hunk 4 above)
Comment (safesparrow) [reply tests/ParallelTypeCheckingTests/Code/TrieMapping.fs:96]: I think we're talking about two very different approaches as the same one - see my general comment below, this probably needs a better chat next week to avoid further misunderstanding (I'm not suggesting the current solution is wrong) 👍

File: tests/ParallelTypeCheckingTests/Code/TrieMapping.fs (hunk 3 above)
Comment (nojaf) [reply tests/ParallelTypeCheckingTests/Code/TrieMapping.fs:57]: - You cannot nest a namespace in F#, so that is why the code looks different for top-level module/namespace than the nested modules. A (top-level) module will always lead to a node that exposes the current file. A top-level namespace does not necessarily expose the current file. Only if it has types or auto-open nested modules.
- `mkTrieForSynModuleDecl` will recursively do detection on the nested modules, it does indeed not take the `[<AutoOpen>]` into account. I believe the reasoning here was that you will have established the link anyway via its parent:
 
image

So, I think at the deeper nested level, though we don't capture it correctly in our trie, I don't think this gap leads to any real problems. I could be wrong here, so don't be shy to come up with scenarios where we don't meet the expectations.

**Comment by nojaf** [comment]
> I think there is a more general issue we're having, which is communication

It appears so, although I think we can easily overcome that problem with a chat.

> For example, I was always convinced that we would end up with a solution that avoids a link if the only reference is an open A.B statement and A.B only contains nested modules.

This is somewhat the case. Consider:
A.fs
```fsharp
namespace X.Y

module A =
   let a = 0
```
B.fs
```fsharp
namespace X.Y

type B = int
```
C.fs
```fsharp
module C

open X.Y
```

Only `B` will be a dependency of `C`. `A` is not exposing itself at the `X` or `Y` level of the trie.

However, if we don't have file B:
A.fs
```fsharp
namespace X.Y

module A =
   let a = 0
```
C.fs
```fsharp
module C

open X.Y
```
`A` will be a dependency of `C` because the `open X.Y` needs to lead to something in the trie.
This case is covered by the `ghost dependency` resolution. After all the content of `C.fs` was processed, we verify that the state does not have any lingering open statements. `open X.Y` will not lead to any files in the trie (as `A` doesn't expose anything at the `Y` level).
To overcome this we look in the child nodes of `Y` if we can find any index that is higher than the index of `C` and we pick this index as a dependency to satisfy the `open X.Y` statement in `C`.

> Or does the code throw when trying to merge two modules with the same name, from different files?
No idea what happens in that scenario 🙈. It will probably throw.

**Comment by nojaf** [comment]
I think this is good to go. Please take a peek at all this.
You would need to approve it before I can merge it.

File: tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs [hunk 5]
```diff
@@ -58,3 +58,256 @@ let ``Basic trie`` () =
...
+    Assert.AreEqual(Set.empty, trie.Files)
+    let xNode = trie.Children.["X"]
+    Assert.AreEqual(Set.empty, xNode.Files)
+    let yNode = xNode.Children.["Y"]
+    Assert.AreEqual(set [| 0 |], yNode.Files)
+    let zNode = yNode.Children.["Z"]
+    Assert.AreEqual(set [| 0 |], zNode.Files)
+
+[<Test>]
+let ``Two modules with the same name, only the first file exposes the index`` () =
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Tests/TrieMappingTests.fs:279]: We should probably take note of this and mention it on the final PR

**Review by safesparrow — APPROVED** [review]
Thanks for the work, I'm happy to merge this.

As a general thought, I think we need to be careful about how we change the results when compilation does not succeed - this is probably the main area our change is going to affect (say the aforementioned reference to a missing nested module).

At some point we should make it clear what the effects are and either convince the F# team that that's fine, or have to change it.
